src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4824,4 +4787,92 @@ private int getClassAccessFlagsRaw() {
+
+    private Method findMethod(boolean publicOnly, String name, Class<?>... parameterTypes) {
+        PublicMethods.MethodList res = getMethodsRecursive(name, parameterTypes, true, publicOnly);
+        return res == null ? null : getReflectionFactory().copyMethod(res.getMostSpecific());
+    }
+
+    /**
+     * Return the first method that meets the requirements of an application main method
+     * {@jls 12.1.4}. The method must:
+     * <ul>
+     * <li>be declared in this class's hierarchy</li>
+     * <li>have the name "main"</li>
+     * <li>have a single argument of type {@code String[]} or no argument</li>
+     * <li>have the return type of void</li>
+     * <li>be public, protected or package private</li>
+     * <li>not be abstract</li>
+     *</ul>
+     *
+     * Searching continues until a main method is found or the search is exhausted. The
+     * primary search occurs in two phases, once for a main method with a {@code
+     * String[]} argument and failing that, once for a main method with a no arguments.
+     * The search itself uses recursion to first look at methods in this class, then
+     * default methods in this class's interface hierarchy and then repeating these steps
+     * with the class's super class.
+     *
+     * @apiNote The method returned may be declared in this class, a super class
+     * or as a default method of an interface that the class or super class
+     * implements. It is not possible to declare a static main method and instance main
+     * method with the same signature in the same class. {@jls 8.4.2} states that
+     * "It is a compile-time error to declare two methods with override-equivalent
+     * signatures in a class."
+     *
+     * @return the main method or null if none found
+     * @throws  SecurityException
+     *          If a security manager, <i>s</i>, is present and any of the
+     *          following conditions is met:
+     *          <ul>
+     *          <li> the caller's class loader is not the same as the
+     *          class loader of this class and invocation of
+     *          {@link SecurityManager#checkPermission
+     *          s.checkPermission} method with
+     *          {@code RuntimePermission("accessDeclaredMembers")}
+     *          denies access to the declared methods within this class
+     *          <li> the caller's class loader is not the same as or an
+     *          ancestor of the class loader for the current class and
+     *          invocation of {@link SecurityManager#checkPackageAccess
+     *          s.checkPackageAccess()} denies access to the package
+     *          of this class
+     *          </ul>
+     * @jls 8.2 Class Members
+     * @jls 8.4 Method Declarations
+     * @jls 8.4.2 Method Signature
+     * @jls 12.1.4 Invoke a main method
+     * @since 22
+     */
+    @PreviewFeature(feature=PreviewFeature.Feature.IMPLICIT_CLASSES)
+    @CallerSensitive
+    public Method findMainMethod() {
```

Comment:
Renaming it to findMainMethod separates it from the existing getXXXMethod methods (good) but it makes it more obvious that this is a method for launchers. I'm not 100% sure that java.lang.Class is the right place for this, it feels out of place.